TU20Bot/Commands/NameMatch.cs

```diff
@@ -0,0 +1,154 @@
+            {"Muhammad Ali","Syed"},
+        };
+
+        [Command("lsuser-assignrole")]
+        [Summary("matches guild member names with a specified list and assigns roles to the matched names")]
+        public async Task matchNames(ulong roleId) {
+            var users = Context.Guild.Users;
+            string discordMessage = "";
+            string noLastNameMsg = "";
+            string noFirstNameMsg = "";
+            string notInGuildMsg = "";
+            for (int i = 0; i < origNames.GetLength(0); i++) {
+                foreach (var user in users) {
+                    string fullName = user.Nickname;
+                    if (fullName == null) {
+                        fullName = user.ToString();
+                        fullName = fullName.Substring(0, fullName.LastIndexOf('#'));
+                    }
+
+                    int spaceIndex = fullName.LastIndexOf(' ');
+                    // If a user doesn't have a first or last name in the server
+                    if (spaceIndex <= 3) {
+                        if (compare((origNames[i, 0] + origNames[i, 1]), fullName)) {
+                            noFirstNameMsg +=
+                                $" User `{origNames[i, 0]} {origNames[i, 1]}` has either " +
+                                $"first or last name which matches with `{fullName}`\n";
+                            Console.WriteLine($" User `{origNames[i, 0]} {origNames[i, 1]}` has either " +
+                                              $"first or last name which matches with `{fullName}`");
+                        }
+                    }
+                    /*
+                     * If a user has both first and last names in the server
+                     * If last name matches then proceed to see if the first name matches
+                     * If both of them match then give the user the role passed in the command
+                     * If first name doesn't match then notify in the console that the last name 
+                     * match was successful.
+                     * If last name doesn't match then notify the user in the console. 
+                     */
+                    else {
+                        string firstName = fullName.Substring(0, spaceIndex);
+                        string lastName = fullName.Substring(spaceIndex + 1);
+
+                        // Checking if the last names match
+                        if (compare(origNames[i, 1], lastName) ||
+                            compare(lastName, origNames[i, 1])) {
+                            // Checking if the first names match
+                            if (compare(origNames[i, 0], firstName) ||
+                                compare(firstName, origNames[i, 0])) {
+                                // Assigning user a role specified in the command
+                                var role = Context.Guild.GetRole(roleId);
+                                await (user as IGuildUser).AddRoleAsync(role);
```

Comment:
Erm... Rider will yell at me for this even though its unlikely user won't be an IGuildUser. Can you use C-style cast? It will make sure a CastException is thrown instead of a NullPointer exception in the worst case scenario (world is ending).
```suggestion
                                await ((IGuildUser) user).AddRoleAsync(role);
```